File: code/R/ch1.R
# Chapter 1 - Exploratory Data Analysis

## libraries ----
library(tidyverse)
library(vioplot)
library(corrplot)
library(gmodels)
library(matrixStats)
library(hexbin)

## import & view data ----
state <- read.csv('data/state.csv')
dfw <- read.csv('data/dfw_airline.csv')
sp500_px <- read.csv('data/sp500_data.csv.gz')
sp500_sym <- read.csv('data/sp500_sectors.csv')
kc_tax <- read.csv('data/kc_tax.csv.gz')


## analysis ----

### location estimates of population and murder rates by state
mean(state[['Population']]) # average population
mean(state[['Population']], trim = 0.1) # trimmed average 
median(state[['Population']]) # median population

weighted.mean(state[['Murder.Rate']], w = state[['Population']])
weightedMedian(state[['Murder.Rate']], w = state[['Population']]) # matrixStats() function

### variability estimates of State population
sd(state[['Population']]) # standard deviation
IQR(state[['Population']]) # interquartile range
mad(state[['Population']]) # median absolute deviation

###  percentiles
quantile(state[['Murder.Rate']], p = c(0.05, 0.25, 0.50, 0.75, 0.95))

### boxplot
boxplot(state[['Population']]/1000000, ylab = 'Population (millions)')

### frequency tables and histograms
breaks <- seq(from = min(state[['Population']]),
              to = max(state[['Population']]), length = 11)
pop_freq <- cut(state[['Population']], breaks = breaks,
                right = TRUE, include.lowest = TRUE)
table(pop_freq)

### histogram
hist(state[['Population']], breaks = breaks)

### density plots and estimates
hist(state[['Murder.Rate']], freq = FALSE)
lines(density(state[['Murder.Rate']]), lwd = 3, col = 'blue')

### exploring binary and categorical data
barplot(as.matrix(dfw) / 6, cex.axis = 0.8, cex.names = 0.7,
        xlab = 'Cause of Delay', ylab = 'Count')

### correlation
etfs <- sp500_px[row.names(sp500_px) > '2012-07-01',
                 sp500_sym[sp500_sym$sector == 'etf', 'symbol']]

corrplot(cor(etfs), method = 'ellipse')

### scatterplot
telecom <- sp500_px[, sp500_sym[sp500_sym$sector == 'telecommunications_services', 'symbol']]
telecom <- telecom[row.names(telecom) > '2012-07-01',]

plot(telecom$T, telecom$VZ, xlab = 'ATT (T)', ylab = 'Verizon (VZ)')

## multivariate analysis: exploring > 2 variables
### hexagonal binning / contours (numeric vs categorical)
kc_tax0 <- subset(kc_tax, TaxAssessedValue < 750000 &
                    SqFtTotLiving > 100 & 
                    SqFtTotLiving < 3500)
nrow(kc_tax0)


ggplot(kc_tax0, (aes(x = SqFtTotLiving, y = TaxAssessedValue))) +
  stat_binhex(color = 'white') +
  theme_bw() +
  scale_fill_gradient(low = 'white', high = 'black') +
  labs(x = 'Finished Square Ft', y = 'Tax-Assessed Value')

ggplot(kc_tax0, aes(SqFtTotLiving, TaxAssessedValue)) +
  theme_bw() +
  geom_point(alpha = 0.1) +
  geom_density2d(color = 'white') +
  labs(x = 'Finished Sq Ft', y = 'Tax-Assessed Value')

## two categorical values
library(descr)

### load data
lc_loans <- read.csv('data/lc_loans.csv')

### contigency table
x_tab <- CrossTable(lc_loans$grade, lc_loans$status,
                    prop.c = FALSE, prop.chisq = FALSE, prop.t = FALSE)

## categorical and numeric data
airline_stats <- read.csv('data/airline_stats.csv')

## compare how % of delays varies across airlines
boxplot(pct_carrier_delay ~ airline, data = airline_stats,
        ylim = c(0, 50))

ggplot(data = airline_stats, aes(airline, pct_carrier_delay)) +
  ylim(0, 50) +
  geom_violin() +
  labs(x = '', y = 'Daily % of Delayed Flights') +
  theme_bw()

## exploring two or more variables

### king county tax eg
kc_tax0 <- subset(kc_tax, TaxAssessedValue < 750000 &
                    SqFtTotLiving > 100 & 
                    SqFtTotLiving < 3500)

ggplot(subset(kc_tax0, ZipCode %in% c(98188, 98105, 98108, 98126)),
       aes(x = SqFtTotLiving, y = TaxAssessedValue)) +
  stat_binhex(color = 'white') +
  theme_bw() +
  scale_fill_gradient(low = 'white', high = 'blue') +
  labs(x = 'Finished Sq Ft', y = 'Tax Assessed Value') +
  facet_wrap('ZipCode')

File: code/python/ch1.py
# Chapter 1 - Exploratory Data Analysis

## packages ----
%matplotlib inline

from pathlib import Path

import pandas as pd
import numpy as np
from scipy.stats import trim_mean
from statsmodels import robust

import seaborn as sns
import matplotlib.pylab as plt

import wquantiles

import os
# os.getcwd()
# os.chdir('')

## estimates of location ----

## import data
state = pd.read_csv('data/state.csv')

dfw = pd.read_csv('data/dfw_airline.csv')

sp500_px = pd.read_csv('data/sp500_data.csv.gz', index_col=0)
sp500_sym = pd.read_csv('data/sp500_sectors.csv')

state['Population'].mean()
trim_mean(state['Population'], 0.1)
state['Population'].median()
print(np.average(state['Murder.Rate'], weights=state['Population']))
print(wquantiles.median(state['Murder.Rate'], weights=state['Population']))

## estimates of variability ----
state['Population'].std() # standard deviation
state['Population'].quantile(0.75) - state['Population'].quantile(0.25) # IQR
robust.scale.mad(state['Population']) # median absolute deviation

## exploring data distribution ----
### percentiles
state['Murder.Rate'].quantile([0.05, 0.25, 0.50, 0.75, 0.95])

### boxplot
ax = (state['Population']/1_000_000).plot.box()
ax.set_ylabel('Population (millions)')

## frequency tables and histograms ----
binnedPopulation = pd.cut(state['Population'], 10)
binnedPopulation.value_counts()

### histogram
ax = (state['Population'] / 1_000_000).plot.hist(figsize = (4, 4))
ax.set_xlabel('Population (millions)')

### density plots and estimates
ax = state['Murder.Rate'].plot.hist(density = True, xlim = [0,12], bins = range(1,12))
state['Murder.Rate'].plot.density(ax = ax)
ax.set_xlabel('Murder Rate (per 100,000)')

## exploring binary and categorical Data
ax = dfw.transpose().plot.bar(figsize = (4, 4), legend = False)
ax.set_xlabel('Cause of Delay')
ax.set_ylabel('Count')

## correlation
etfs = sp500_px.loc[sp500_px.index > '2012-07-01',
sp500_sym[sp500_sym['sector'] == 'etf']['symbol']]

sns.heatmap(etfs.corr(), vmin = -1, vmax = 1,
cmap = sns.diverging_palette(20, 220, as_cmap = True))

## scatterplot
# Determine telecommunications symbols
telecomSymbols = sp500_sym[sp500_sym['sector'] == 'telecommunications_services']['symbol']
telecom = sp500_px.loc[sp500_px.index >= '2012-07-01', telecomSymbols]


ax = telecom.plot.scatter(x = 'T', y = 'VZ', figsize = (4, 4), marker = '$\u25EF$')
ax.set_xlabel('ATT (T)')
ax.set_ylabel('Verizon (VZ)')
ax.axhline(0, color = 'grey', lw = 1)
ax.axvline(0, color = 'grey', lw = 1)

## multivariate Analysis
kc_tax0 = kc_tax.loc[(kc_tax.TaxAssessedValue < 750000) &
                     (kc_tax.SqFtTotLiving > 100) &
                     (kc_tax.SqFtTotLiving < 3500), :]
kc_tax0.shape

### hex binning
ax = kc_tax0.plot.hexbin(x = 'SqFtTotLiving', y = 'TaxAssessedValue',
gridsize = 30, sharex = False, figsize = (5, 4))
ax.set_xlabel('Finished Square Feet')
ax.set_ylabel('Tax-Assessed Value')

### density
fig, ax = plt.subplots(figsize=(4, 4))
sns.kdeplot(data=kc_tax0.sample(10000), x='SqFtTotLiving', y='TaxAssessedValue', ax=ax)
ax.set_xlabel('Finished Square Feet')
ax.set_ylabel('Tax Assessed Value')
